KAN-242-알림 기능 관련 클래스 및 메서드명 수정, 패키지 분리

backswan0 opened the pull request:
## #️⃣ CHEERHA Board Number

https://spring-3-jo.atlassian.net/browse/KAN-242

## 📝 요약

- '알림'이라는 의미가 잘 드러나도록 엔티티 클래스 이름을 'Notification'으로 수정
- 클래스, 변수, 메서드 이름을 전반적으로 수정 
- UserKeyword 삭제 시 Request Body 대신 Request Param을 사용하도록 수정 
- 패키지 분리 
- 사용자와 채용 공고를 매칭하는 로직을 각각 private 메서드로 추출하여 흐름이 한눈에 보이도록 리팩토링

## 🛠️ PR 유형

- [ ] 배포용 PR : dev 테스트를 완료하였나요?
- [ ] 새로운 기능 추가
- [ ] 버그 수정
- [x] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [x] 코드 리팩토링
- [x] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## 💬 논의사항 or 질문

## ✅ PR Checklist

PR이 다음 요구 사항을 충족했는지 확인하기!

- [x] 커밋 메시지 컨벤션 준수
- [x] 변경 사항 테스트 여부 체크(버그 수정/기능 테스트)

## 💥 리뷰어 지목!!!!!

<!--- 리뷰어를 2명 이상 지목해주세요. -->
- [x] 채영
- [ ] 지현
- [x] 리은
- [x] 승찬
- [ ] 주양

backswan0 committed:
:memo: chore: 이름 수정

- 불필요한 레이어는 주석 처리, 이후 삭제 예정

- 클래스 및 메서드 이름이 '알림' 관련임이 잘 드러나도록 다시 명명

backswan0 committed:
Merge remote-tracking branch 'origin/dev' into refactor/email-customizing

backswan0 committed:
:memo: chore: 주석 해제

- 테스트 때문에 주석 처리한 '이메일 알림 활성화 된 사람만 조회'하는 쿼리문 주석 해제

backswan0 committed:
:fire: delete: 알림 생성용 데이터 조회 레이어 삭제

- 스케줄러와 하는 일이 겹쳐서 불필요하다고 판단하여 삭제함

backswan0 committed:
:recycle: refactor: 알림 생성 서비스 메서드 분리

- 특수 로직인 만큼 별도의 클래스 생성 대신 private 메서드로 추출

- [할 일] 줄 바꿈 및 주석 작성

backswan0 committed:
:memo: chore: 메서드 위치 변경

- 로직이 잘 보이도록 지나치게 긴 private 메서드를 맨 밑으로 위치 변경

- 메서드명 수정

backswan0 committed:
:recycle: refactor: Delete 메서드 호출 시 Request Body 대신 Request Param 사용

- Request Body 대신 Request Param으로 사용자가 선택한 채용 키워드 목록을 지울 수 있도록 리팩토링

- 사용하지 않는 DeleteUserKeywordRequestDto 삭제

backswan0 committed:
:memo: chore: 이름 수정

- 알림을 받는 사람의 정보라는 의미로 'UserDto'를 'NotificationRecipientDto'로 변경

- 패키지 이름 또한 'notice'에서 'notification'으로 변경

backswan0 committed:
:memo: chore: 주석 해제

- 테스트용으로 주석 처리한 부분을 다시 해제함

backswan0 committed:
:memo: chore: 주석 작성 및 불필요한 파일 삭제

- 각 메서드에 주석 작성

- 사용하지 않는 JobOpeningKeywordController 파일 삭제

roqkfchqh commented on src/main/java/com/project/cheerha/domain/notification/sender/EmailSender.java:
```diff
@@ -24,110 +24,120 @@
+     * @param recipientEmail  수신자 이메일
+     * @param notificationSet 해당 수신자에게 보낼 알림 목록
+     */
+    private void sendMail(String recipientEmail, Set<Notification> notificationSet) {
```
이건 notification용 메일 템플릿이니까, 따로 다른 클래스로 빼는 게 좋아보입니다 !

roqkfchqh replied:
![스크린샷 2025-03-05 오후 12 01 14](https://github.com/user-attachments/assets/51993dde-42d6-4252-89ac-fa41e7f5a4d8)
최종적으로 이런 형태면 좋을 것 같아용
notification과 verification format을 분리한 뒤에,
sendGrid외에 다른 이메일 플랫폼으로 갈아낄 수 있도록 EmailSender 인터페이스를 만들고
해당 인터페이스를 상속받는 sendGrid 콘크리트를 만들어가지고 관리하는거져

roqkfchqh replied:
![스크린샷 2025-03-05 오후 12 02 14](https://github.com/user-attachments/assets/1e3050d0-ceab-4bd1-9554-8b55fe705437)
![스크린샷 2025-03-05 오후 12 02 20](https://github.com/user-attachments/assets/a96c1137-c769-4703-b0ae-ee07a89e2b3f)
![스크린샷 2025-03-05 오후 12 02 31](https://github.com/user-attachments/assets/af10ac27-be6a-4102-8d29-bd46ad4bf997)
![스크린샷 2025-03-05 오후 12 02 44](https://github.com/user-attachments/assets/6c10d3ee-aeff-418b-b1d5-da79775bc95f)
![스크린샷 2025-03-05 오후 12 02 50](https://github.com/user-attachments/assets/842de706-05ff-4daf-9dfd-d4d998cc6181)
대충리팩토링한거라 컴파일에러는 해결 못했는데 암튼 이런느낌이면 책임분리 명확히 되어용

roqkfchqh replied:
추가로 common/email로 패키지 옮기고, notification 패키지에는 notificationEmailSender만 남도록 하면 더 깔끔할 듯 합니다..!
(다른 서비스에서 재사용 가능하니까요!)
common/email/format 까지도 나눌 수 있어용.

roqkfchqh commented on src/main/java/com/project/cheerha/domain/notification/sender/EmailSender.java:
```diff
@@ -24,110 +24,120 @@
+     * @param recipientEmail 이메일 수신자
+     * @param code           인증 코드
+     */
+    public void sendVerificationEmail(String recipientEmail, String code) {
```
이것도 같이 빼주세영(다른 클래스로)

roqkfchqh replied:
리팩토링 예시 댓글 참고부탁드려용 (거기 관련내용 다 적어놨어요)

roqkfchqh commented on src/main/java/com/project/cheerha/domain/notification/sender/EmailSender.java:
```diff
@@ -24,110 +24,120 @@
 
+    // Notification을 이메일로 비동기 전송
     @Async
     public void sendEmails() {
```
무슨 이메일인지 정확히 명시 부탁드립니다

roqkfchqh replied:
리팩토링 예시 댓글 참고부탁드려용 (거기 관련내용 다 적어놨어요)

tmdcksdl commented on src/main/java/com/project/cheerha/domain/notification/scheduler/NotificationDataFetchingScheduler.java:
```diff
@@ -0,0 +1,40 @@
+public class NotificationDataFetchingScheduler {
+
+    private final NotificationService notificationService;
+    private final NotificationDataProviderQuery notificationDataProviderQuery;
```
현재 NotificationDataFetchingScheduler에서 NotificationDataProviderQuery를 별도로 주입받고 있는데,
NotificationRepository가 NotificationDataProviderQuery를 extends하면 하나의 Repository로 통합할 수 있어요!

![image](https://github.com/user-attachments/assets/19fdf1a0-d141-42e6-9a57-63314694ec81)

이런 식으로 변경해서 NotificationRepository를 주입받는 방법은 어떨까요?!

backswan0 committed:
Merge remote-tracking branch 'origin/dev' into refactor/email-customizing

backswan0 committed:
:recycle: refactor: 이메일 전송 로직 분리 및 인터페이스와 구현체 클래스 생성

- 이메일 전송 인터페이스 및 구현체 클래스 생성

backswan0 committed:
🎉 add: 이메일 양식 전용 파일 생성

- 이메일로 채용 공고 목록을 전달해줄 때 사용할 양식 클래스 생성 및 기존 파일에서 해당 로직 분리

- 이메일로 인증 코드 전송할 때 사용할 양식 클래스 생성 및 기존 파일에서 해당 로직 분리

backswan0 committed:
:tada: add: 이메일 종류에 따라 전송 클래스 생성

- 이메일 알림만 보내는 로직을 기존 파일에서 분리 및 별도로 클래스 생성

- 인증용 이메일만 보내는 로직을 기존 파일에서 분리 및 별도로 클래스 생성

backswan0 committed:
:recycle: refactor: 메서드 대체

- 분리한 로직에 맞추어 메서드 대체

backswan0 committed:
:memo: chore: 주석 처리

- 사용하지 않는 기존 파일 주석 처리

- [할 일] 기록 후 삭제

backswan0 committed:
Merge remote-tracking branch 'origin/dev' into refactor/email-customizing

backswan0 committed:
:truck: move: 파일 이동

- common 패키지에 이메일 양식 및 SendGrid 전송 클래스와 이메일 인증 전송 클래스 이동

- notification 패키지에는 이메일 알림 전송 관련 파일만 남김

backswan0 committed:
:truck: move: 파일 이동

- NotificationEmailSender를 common/email/sender 패키지로 이동

- 사용하지 않는 import 삭제

roqkfchqh commented on src/main/java/com/project/cheerha/common/email/sender/EmailSender.java:
```diff
@@ -0,0 +1,7 @@
+import java.io.IOException;
+
+public interface EmailSender {
+    void sendEmailBySendGrid(String recipientEmail, String subject, String content) throws IOException;
```
인터페이스의 메서드명은 가장 추상적인 형태로 둬야합니다! 예를들면 send로용

backswan0 replied:
넵, 수정하겠습니다:)

roqkfchqh commented on src/main/java/com/project/cheerha/common/email/sender/NotificationEmailSender.java:
```diff
@@ -0,0 +1,63 @@
+            String content = emailData[1];
+
+            // 이메일 전송
+            sendGridEmailSender.sendEmailBySendGrid(recipientEmail, subject, content);
```
인터페이스를 호출해야 결합도가 낮아집니당

backswan0 replied:
제가 인터페이스를 호출해놓고 이름을 구현체로 적었네요; 수정하겠습니다!

roqkfchqh commented on src/main/java/com/project/cheerha/common/email/sender/VerificationEmailSender.java:
```diff
@@ -0,0 +1,29 @@
+            String subject = emailData[0];
+            String content = emailData[1];
+
+            sendGridEmailSender.sendEmailBySendGrid(recipientEmail, subject, content);
```
여기서두요! 왜냐면 다른 걸로 갈아끼기가 쉬워져요(샌드그리드에서 구글로 메일플랫폼이 변해도 이 코드엔 변화가 없을거예요)

backswan0 replied:
제가 인터페이스를 호출해놓고 이름을 구현체로 적었네요; 수정하겠습니다! 피드백 감사합니다!:)

backswan0 committed:
:memo: chore: 메서드 및 의존성 이름 수정

- 구현체 이름으르 적은 부분 인터페이스 클래스 이름으로 수정

- 인터페이스의 메서드 이름을 가장 추상적인 형태인 'send()'로 수정
